engine/src/main/java/io/camunda/zeebe/engine/state/mutable/MutableUserTaskState.java

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.engine.state.mutable;
+
+import io.camunda.zeebe.engine.state.immutable.UserTaskState;
+import io.camunda.zeebe.protocol.impl.record.value.usertask.UserTaskRecord;
+
+public interface MutableUserTaskState extends UserTaskState {
+
+  void create(final UserTaskRecord userTask);
+
+  void update(final UserTaskRecord userTask);
+
+  void delete(final UserTaskRecord userTask);
```

Comment:
:wrench: @tmetzke I don't think we should remove the `userTaskKey` from the input arguments here. (also referring to what you wrote [here](https://github.com/camunda/zeebe/pull/15182#issuecomment-1813995012))

This is for two reasons:
1. It would differ from the pattern the other State classes apply.
2. The ZPA team tries to explicitly define what is used in a method as input arguments of that method. Feel free to raise this topic with other team members.

Reply:
@korthout I know you usually have a preference on how we define input arguments for methods. Let me know if what I wrote here isn't true.

Reply:
Thanks, @koevskinikola, makes a lot of sense to me in general. I actually got the idea of providing only records from the [MutableFormState](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/state/mutable/MutableFormState.java) class. We do the same over there. For the job state class, that won't be possible since it doesn't carry along the key itself. If we add the `userTaskKey` to the method parameters, do we ensure in those methods that the provided `userTaskKey` parameter and the one on the provided record are the same?

In general, I also am totally fine with keeping the `userTaskKey` in the methods, to be honest. No hard feelings here.

Reply:
My personal preference is to pass only the key in the case of deletion because:
- it needs to be clarified what the expected behavior of delete is when passing a UserTaskRecord object. If some of the properties don't match the data in the state, should the deletion be ignored because it is not the same user task, or should it still be considered the same user task because the key is equivalent? This is unclear from the name and method schema so it would require additional JavaDoc. We can avoid this simply by cleaning up the method schema.
- it's easier to test the method when you only need to provide the bare minimum of data to call it
- using a parameter object can be a helpful refactoring when the number of method parameters exceeds a point where the method calls become hard to read. There is no need for this when there is just a single parameter.

Reply:
Thanks, @korthout, for the detailed feedback! ❤️ 

Some thoughts and questions regarding this:
* I totally agree regarding the deletion method! Excellent catch, good argumentation 👍 
* How about the other methods (create and update)? I'd argue that passing the key additionally makes the method less clear, similar to the reasons you mentioned for the deletion. Should the creation/update be rejected or ignored because the key and the `userTaskKey` in the record don't match? Should we silently update the `userTaskKey` in the provided record? If we keep it to the record only, we only store what you provide with the record.

What do you think, @korthout?

Reply:
>How about the other methods (create and update)

Typically, the record key is record metadata and not included in the record value. However, for clarity we sometimes include the record key as a property as well. Although not entirely necessary, this can be useful for any consumer of the records (e.g. API responses, debugging, exporter consumers, etc). But we don't do this for all records.

As you show in the questions you raise, double bookkeeping leads to strange questions when you can't keep the duplicate values in sync. That becomes visible when you want to program defensively.

IMO, it's fine to go either direction, they both sound reasonable:
- keep it aligned with places where we don't duplicate the data inside the record value
- keep it simple when the data is duplicated inside the record value

Reply:
Thanks again, @korthout, that resonates well with me 👍

I'll give this some more thought. My gut feeling is to use only the record value where the key is part of the record data and the record needs to be passed. If only a small number of parameters is required instead of the record (e.g. for deletion), I'd go with only those parameters (I totally agree with your reasoning). As far as I can tell, the form records are mostly handled like this in our code but could be improved a little bit (e.g. the deletion use case).

Reply:
@tmetzke In the future, we want to move away from these methods and towards a more fine-grained column family access. That would emphasize that changes to state classes often require new versions of event appliers (this is currently opaque and often leads to bugs) https://github.com/camunda/zeebe/issues/13813

Reply:
Thanks, @korthout, that is really interesting to read and follow 👍 
In a second iteration (i.e. with the second increment of the related Zeebe Epic), I will refine this interface to more fine-grained methods referring to ColumnFamilies so it will be easier to turn this into whatever option we agree on with #13813.

Reply:
I adjusted the state API to only use the key for deletion and use records-only for the other. We'll refine the API with the next increments. Closing here for now. Thanks again, @koevskinikola and @korthout for the great feedback! 🚀